### PR TITLE
[NVIDIA TF] Enable BF16 MatMul Ops

### PR DESCRIPTION
**Pull request by kaixih**
This PR enables BF16 MatMul/BatchedMatMul via cuBLAS.

cc. @nluehr @pjannaty @kushanam @trevor-m

### CODE REVIEW

**Comment by kaixih** [comment]
I will remove the `TF_ENABLE_GPU_BFLOAT16` flag and fix some tolerance issues in the unit tests as pointed out by @reedwm in PR https://github.com/tensorflow/tensorflow/pull/57674.

**Review by reedwm — CHANGES_REQUESTED** [review]

**Review by reedwm — CHANGES_REQUESTED** [review]

File: tensorflow/python/kernel_tests/math_ops/batch_matmul_op_test.py [hunk 1]
```diff
@@ -29,9 +30,11 @@
 
 
 def GetRandomNormalInput(shape, dtype):
-  # float16 has limited range so we reduce the variance of the scalars.
-  scale = 10.0 if dtype != np.float16 else 0.1
-  loc = -10.0 if dtype != np.float16 else 0.1
+  # float16 or bfloat16 has limited range so we reduce the variance of the
+  # scalars.
```
[the remark this answers is not in the document]
Comment (kaixih) [reply tensorflow/python/kernel_tests/math_ops/batch_matmul_op_test.py:34]: Oh, yes, I think what I meant to do was to limit the scale of the inputs, otherwise, I saw the bfloat16 will output like:
```
Not equal to tolerance rtol=19.8425, atol=19.8425
Mismatched value: a is different from b.
not close where = (array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), array([ 60,  62,  65,  66,  67,  68,  69, 120, 121, 122, 123, 124, 125,
       126, 127, 128]))
not close lhs = [-25.    -25.    -25.    -25.    -25.    -25.    -25.    -22.375 -22.375
 -22.375 -22.375 -22.375 -22.375 -22.375 -22.375 -22.375]
not close rhs = [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
not close dif = [25.    25.    25.    25.    25.    25.    25.    22.375 22.375 22.375
 22.375 22.375 22.375 22.375 22.375 22.375]
not close tol = [19.842514 19.842514 19.842514 19.842514 19.842514 19.842514 19.842514
 19.842514 19.842514 19.842514 19.842514 19.842514 19.842514 19.842514
 19.842514 19.842514]
dtype = bfloat16, shape = (1, 210)
Mismatched elements: 16 / 210 (7.62%)
Max absolute difference: 106.625
Max relative difference: 0.82655036
 x: array([[-46.75 , -46.75 , -46.75 , -46.75 , -46.75 , -46.75 , -46.75 ,
        -46.75 , -46.75 , -46.75 , -51.   , -51.   , -51.   , -51.   ,
        -51.   , -51.   , -51.   , -51.   , -51.   , -51.   , -52.5  ,...
 y: array([[-129., -129., -129., -129., -129., -129., -129., -129., -129.,
        -129., -129., -129., -129., -129., -129., -129., -129., -129.,
        -129., -129., -129., -129., -129., -129., -129., -129., -129.,...

```

[the remark this answers is not in the document]
Comment (reedwm) [reply tensorflow/python/kernel_tests/math_ops/batch_matmul_op_test.py:34]: But why would a large scale result in such incorrect outputs? It shouldn't make a difference whether the scale is .1 or 10.

[the remark this answers is not in the document]
Comment (kaixih) [reply tensorflow/python/kernel_tests/math_ops/batch_matmul_op_test.py]: Yes, I think I found the root cause. It is because the loss function used is the reduce_sum, which would generate high error over these bfloat16 values ranging 10.0. So, I cast the results to float32 before the reduce_sum and keep the inputs as what they were. New commit has been push. PTAL. @reedwm

**Review by reedwm — APPROVED** [review]

**Review by reedwm — APPROVED** [review]

**Comment by kaixih** [comment]
Gentle ping. @reedwm

**Comment by kaixih** [comment]
Noticed a couple of newly added unit tests of bf16 failed on CPU. Just skip these tests on CPU, since they are not focus of this PR. @reedwm

**Review by reedwm — APPROVED** [review]

**Comment by kaixih** [comment]
@reedwm It seems the merging is stuck at some point. Can you help check?
